openedx_user_groups/api.py

```diff
@@ -0,0 +1,418 @@
+
+    This method considers the mutual exclusivity of the groups in the collection.
+
+    Args:
+        group_collection_id (int): The ID of the group collection.
+
+    Returns:
+        tuple:
+        - GroupCollection: The group collection.
+        - QuerySet: The duplicates users found and removed from the group collection.
+    """
+    with transaction.atomic():
+        group_collection = get_group_collection_by_id(group_collection_id)
+        for group in group_collection.user_groups.all():
+            evaluate_and_update_membership_for_group(group.id)
+        # Find duplicates in the group collection to remove them and prompt for action
+        duplicates = (
+            User.objects.filter(
+                usergroupmembership__group__in=group_collection.user_groups.all()
+            )
+            .annotate(group_count=Count("usergroupmembership__group", distinct=True))
+            .filter(group_count__gt=1)
+        )
+        if duplicates.exists():
+            # TODO: Prompt for action, but for the time being remove the duplicates
+            for duplicate in duplicates:
+                duplicate.usergroupmembership_set.filter(
+                    group__in=group_collection.user_groups.all()
+                ).delete()
+        return group_collection, duplicates
+
+
+def get_available_registered_criteria_schema():
+    """Get all available with their schema for fields, operators and descriptions.
+
+    Returns:
+        dict: A dictionary containing the schema for all available criteria. For example:
+        {
+            "criterion_type": {
+                "fields": {
+                    "field_name": {
+                        "type": "string",
+                        "description": "Description of the field"
+                    }
+                },
+                "operators": ["operator1", "operator2"],
+                "description": "Description of the criterion",
+                "criterion_type": "criterion_type",
+                "supported_scopes": ["course", "organization", "instance"]
+            }
+        }
+    """
+    return {
+        criterion_type: criterion_class.get_schema()
+        for criterion_type, criterion_class in CriterionManager.get_criterion_classes().items()
+    }
+
+
+# TODO: THESE METHODS I HAVEN'T TESTED YET
+
+
+def get_group_by_name_and_scope(name: str, scope: str):
+    """Get a user group by its name and scope.
+
+    Args:
+        name (str): The name of the user group.
+        scope (str): The scope of the user group.
+
+    Returns:
+        UserGroup: The user group.
+    """
+    return UserGroup.objects.get(name=name, scope=scope)
+
+
+def get_user_group_members(group_id: int):
+    """Get the members of a user group.
+
+    Args:
+        group_id (int): The ID of the user group.
+
+    Returns:
+        list: A list of users that are members of the user group.
+    """
+    return UserGroup.objects.get(id=group_id).users.all()
```

Comment:
We may need this (or a separate function) to be paginated / a generator or something similar for cases where the group is huge ("all active non-staff users in an instance"?).